### PR TITLE
feat(core): update sign-in-experience

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

- Update Zod jsonb guard of sign-in experiences
- Update sign-in experience API

## Linear Issue Reference

- LOG-1811
- Related
    - LOG-2045: Add more test cases for GET /sign-in-exp

## Testing

image

image

---
@logto-io/eng

### CODE REVIEW

File: packages/core/src/routes/sign-in-experience.test.ts [hunk 1]
```diff
@@ -20,48 +55,91 @@ jest.mock('@/queries/sign-in-experience', () => ({
...
-      slogan: 'silverhand',
-    };
+it('GET /sign-in-exp', async () => {
+  const response = await signInExperienceRequester.get('/sign-in-exp');
+  expect(response.status).toEqual(200);
+  expect(response.body).toEqual(mockSignInExperience);
+});
+
+describe('PATCH /sign-in-exp', () => {
+  describe('should fail without enabled connector when enabled sign-in', () => {
```
Comment (wangsijie) [inline packages/core/src/routes/sign-in-experience.test.ts:74]: nit: this sentence is kind of wired.

File: packages/core/src/routes/sign-in-experience.test.ts [hunk 2]
```diff
@@ -20,48 +55,91 @@ jest.mock('@/queries/sign-in-experience', () => ({
...
+    beforeEach(() => {
+      getConnectorInstances.mockImplementationOnce(async () => []);
```
Comment (wangsijie) [inline packages/core/src/routes/sign-in-experience.test.ts:76]: nit: mockResolvedValue([])

File: packages/core/src/routes/sign-in-experience.ts [hunk 3]
```diff
@@ -41,11 +52,48 @@ export default function signInExperiencesRoutes<T extends AuthedRouter>(router:
...
     async (ctx, next) => {
-      const { body } = ctx.guard;
+      const { socialSignInConnectorIds, ...rest } = ctx.guard.body;
+      const { signInMethods } = rest;
+
+      if (signInMethods) {
+        const connectorInstances = await getConnectorInstances();
+
+        if (signInMethods.email !== SignInMethodState.disabled) {
+          checkEnabledConnectorByType(connectorInstances, ConnectorType.Email);
```
Comment (wangsijie) [inline packages/core/src/routes/sign-in-experience.ts:62]: can we replace this function with an `assertThat`?

File: packages/schemas/src/foundations/jsonb-types.ts [hunk 4]
```diff
@@ -87,23 +89,51 @@ export enum BrandingStyle {
...
-  darkBackgroundColor: z.string(),
-  style: z.nativeEnum(BrandingStyle),
-  logoUrl: z.string().optional(),
-  slogan: z.string().optional(),
-});
+export const brandingGuard = z
+  .object({
+    primaryColor: z
+      .string()
+      .nonempty()
```
[the remark this answers is not in the document]
Comment (IceHe) [reply packages/schemas/src/foundations/jsonb-types.ts:96]: **seed** (TODO) will fill the default colors.

when updating, required to provide the whole jsonb field.

File: packages/core/src/routes/sign-in-experience.ts (hunk 3 above)
Comment (IceHe) [reply packages/core/src/routes/sign-in-experience.ts]: refactored, PTAL
```ts
const assertEnabledConnectorByType = (
  connectorInstances: ConnectorInstance[],
  type: ConnectorType
) => {
  assertThat(
    connectorInstances.some((item) => item.connector.enabled && item.metadata.type === type),
    'sign_in_experiences.enabled_connector_not_found',
    { type }
  );
};
```

Comment (IceHe) [reply packages/core/src/routes/sign-in-experience.ts]: also replaced other `if(){ throw }` with `assertThat`, PTAL

File: packages/core/src/routes/sign-in-experience.test.ts (hunk 1 above)
Comment (IceHe) [reply packages/core/src/routes/sign-in-experience.test.ts]: Updated, PTAL

`should fail when enabled sign-in and there is no enabled connector`

File: packages/schemas/src/foundations/jsonb-types.ts [hunk 5]
```diff
@@ -1,5 +1,7 @@
 import { z } from 'zod';
 
+const isValidHexColor = (color: string) => /^#[\da-f]{3}([\da-f]{3})?$/i.test(color);
```
[the remark this answers is not in the document]
Comment (IceHe) [reply packages/schemas/src/foundations/jsonb-types.ts:3]: `#000` is also valid hex color.

File: packages/core/src/routes/sign-in-experience.ts [hunk 6]
```diff
@@ -41,11 +72,39 @@ export default function signInExperiencesRoutes<T extends AuthedRouter>(router:
...
+          assertThat(
+            socialSignInConnectorIds && socialSignInConnectorIds.length > 0,
+            'sign_in_experiences.empty_social_connectors'
+          );
+
+          assertEnabledSocialConnectorsByIds(socialSignInConnectorIds, connectorInstances);
+        }
+      }
+
+      // Only when the parameter signInMethods.social is enabled, do update socialSignInConnectorIds.
```
Comment (simeng-li) [inline packages/core/src/routes/sign-in-experience.ts:101]: nit: Only update socialSignInConnectorIds when the parameter signInMethods.social is enabled

File: packages/schemas/src/foundations/jsonb-types.ts [hunk 7]
```diff
@@ -87,23 +90,51 @@ export enum BrandingStyle {
...
+export const brandingGuard = z
+  .object({
+    primaryColor: z
+      .string()
+      .nonempty()
+      .refine((color) => isValidHexColor(color), invalidHexColorMessage),
+    backgroundColor: z
+      .string()
+      .nonempty()
+      .refine((color) => isValidHexColor(color), invalidHexColorMessage),
```
Comment (simeng-li) [inline packages/schemas/src/foundations/jsonb-types.ts:102]: Lets discuss whether is a good practice to assert the custom refine validations here.

File: packages/schemas/src/foundations/jsonb-types.ts [hunk 8]
```diff
@@ -87,23 +90,51 @@ export enum BrandingStyle {
...
+    style: z.nativeEnum(BrandingStyle),
+    logoUrl: z.string().url(),
+    slogan: z.string().nonempty().optional(),
+  })
+  .refine((branding) => {
+    if (branding.style === BrandingStyle.Logo_Slogan) {
+      return branding.slogan && branding.slogan.trim() !== '';
+    }
+
+    return true;
```
Comment (simeng-li) [inline packages/schemas/src/foundations/jsonb-types.ts:121]: Especially this one, it is more like a logic validation, zod should only keep clean type/format validation?

File: packages/schemas/src/foundations/jsonb-types.ts [hunk 9]
```diff
@@ -87,23 +90,51 @@ export enum BrandingStyle {
...
-  contentUrl: z.string().optional(),
-});
+export const termsOfUseGuard = z
+  .object({
+    enabled: z.boolean(),
+    contentUrl: z.string().url().optional(),
+  })
+  .refine((termsOfUse) => {
+    if (termsOfUse.enabled) {
+      return termsOfUse.contentUrl;
```
Comment (simeng-li) [inline packages/schemas/src/foundations/jsonb-types.ts:133]: why return termsOfUse.contentUrl; here ?

File: packages/schemas/src/foundations/jsonb-types.ts [hunk 10]
```diff
@@ -126,12 +157,23 @@ export enum SignInMethodState {
...
+  })
+  .refine((signInMethods) => {
+    const signInMethodStates = Object.values(signInMethods);
+    const primarySignInMethodCount = signInMethodStates.reduce(
+      (primaryCount, signInMethodState) =>
+        primaryCount + (signInMethodState === SignInMethodState.primary ? 1 : 0),
+      0
+    );
+
+    return primarySignInMethodCount === 1;
```
Comment (simeng-li) [inline packages/schemas/src/foundations/jsonb-types.ts:175]: Ditto: feels like these are more like logic validations.

**Review by simeng-li — CHANGES_REQUESTED** [review]

File: packages/core/src/routes/sign-in-experience.ts [hunk 11]
```diff
@@ -41,11 +56,51 @@ export default function signInExperiencesRoutes<T extends AuthedRouter>(router:
...
+        if (isEnabled(signInMethods.email)) {
+          assertEnabledConnectorByType(ConnectorType.Email, connectorInstances);
+        }
+
+        if (isEnabled(signInMethods.sms)) {
+          assertEnabledConnectorByType(ConnectorType.SMS, connectorInstances);
+        }
+
+        if (isEnabled(signInMethods.social)) {
+          assertEnabledConnectorByType(ConnectorType.Social, connectorInstances);
```
[the remark this answers is not in the document]
Comment (IceHe) [reply packages/core/src/routes/sign-in-experience.ts:90]: I'm good with both.

We can remove this line, and the only difference is the error message.
image
or
image

@logto-io/eng any good idea?

[the remark this answers is not in the document]
Comment (IceHe) [reply packages/core/src/routes/sign-in-experience.ts:90]: I'm good with both.

File: packages/core/src/routes/sign-in-experience.ts [hunk 12]
```diff
@@ -41,11 +56,51 @@ export default function signInExperiencesRoutes<T extends AuthedRouter>(router:
...
+                  instance.connector.enabled && instance.metadata.type === ConnectorType.Social
+              ),
+            'sign_in_experiences.invalid_social_connectors'
+          );
+        }
+      }
+
+      // Only when the parameter signInMethods.social is enabled, do update socialSignInConnectorIds.
+      const signInExperience =
+        signInMethods && signInMethods.social !== SignInMethodState.disabled
```
[the remark this answers is not in the document]
Comment (IceHe) [reply packages/core/src/routes/sign-in-experience.ts:103]: Tried and eslint disallow:
image

File: packages/core/src/routes/sign-in-experience.ts (hunk 6 above)
Comment (IceHe) [reply packages/core/src/routes/sign-in-experience.ts:101]: Or `Only update socialSignInConnectorIds when social sign-in is enabled`?

Updated, PTAL.

File: packages/schemas/src/foundations/jsonb-types.ts (hunk 9 above)
Comment (IceHe) [reply packages/schemas/src/foundations/jsonb-types.ts:133]: It's just a boolean check.
It's a little confusing.

File: packages/core/src/routes/sign-in-experience.ts [hunk 13]
```diff
@@ -41,11 +56,51 @@ export default function signInExperiencesRoutes<T extends AuthedRouter>(router:
...
+
+          const socialConnectorIdSet = new Set(socialSignInConnectorIds);
+
+          // Should provide enabled social connectors
+          assertThat(
+            connectorInstances
+              .filter((instance) => socialConnectorIdSet.has(instance.connector.id))
+              .every(
+                (instance) =>
+                  instance.connector.enabled && instance.metadata.type === ConnectorType.Social
```
[the remark this answers is not in the document]
Comment (IceHe) [reply packages/core/src/routes/sign-in-experience.ts]: Refactored, PTAL
```ts
          const enabledSocialConnectorIds = new Set(
            enabledConnectorInstances
              .filter((instance) => instance.metadata.type === ConnectorType.Social)
              .map((instance) => instance.connector.id)
          );

          assertThat(
            socialSignInConnectorIds.some((id) => !enabledSocialConnectorIds.has(id)),
            'sign_in_experiences.invalid_social_connectors'
          );
```

- `getEnabledConnectorInstances` gets the necessary connectors from DB once for checking email, SMS, and social connectors.
- Reuse the result of `getEnabledConnectorInstances` to computing enabled social connectors

File: packages/core/src/routes/sign-in-experience.ts [hunk 14]
```diff
@@ -41,11 +56,51 @@ export default function signInExperiencesRoutes<T extends AuthedRouter>(router:
       body: SignInExperiences.createGuard.omit({ id: true }).partial(),
     }),
     async (ctx, next) => {
-      const { body } = ctx.guard;
+      const { socialSignInConnectorIds, ...rest } = ctx.guard.body;
+      const { signInMethods } = rest;
+
+      if (signInMethods) {
+        const connectorInstances = await getConnectorInstances();
```
[the remark this answers is not in the document]
Comment (IceHe) [reply packages/core/src/routes/sign-in-experience.ts]: Refactored, PTAL.

File: packages/core/src/routes/sign-in-experience.test.ts [hunk 15]
```diff
@@ -1,12 +1,47 @@
...
+  connector: {
+    id: 'google',
+    enabled: false,
+    config: {},
+    createdAt: 1_646_382_233_000,
+  },
+  metadata: {
+    type: ConnectorType.Social,
+  },
+};
```
[the remark this answers is not in the document]
Comment (IceHe) [reply packages/core/src/routes/sign-in-experience.test.ts]: There is no need to reuse it in other places.
So keep it for now?

File: packages/schemas/src/foundations/jsonb-types.ts [hunk 16]
```diff
@@ -87,23 +90,51 @@ export enum BrandingStyle {
...
-  style: z.nativeEnum(BrandingStyle),
-  logoUrl: z.string().optional(),
-  slogan: z.string().optional(),
-});
+export const brandingGuard = z
+  .object({
+    primaryColor: z
+      .string()
+      .nonempty()
+      .refine((color) => isValidHexColor(color), invalidHexColorMessage),
```
Comment (darcyYe) [inline packages/schemas/src/foundations/jsonb-types.ts:98]: as we discussed offline, maybe no need to use refine() here and can use string().regex(someRegEx) instead?

**Comment by IceHe** [comment]
Done in separate PRs. Abandoned.

**Pull request closed without merging**